src/Compilers/CSharp/Portable/Symbols/FileIdentifier.cs

```diff
@@ -9,33 +9,91 @@
 
 namespace Microsoft.CodeAnalysis.CSharp;
 
-internal readonly struct FileIdentifier
+internal class FileIdentifier
 {
     private static readonly Encoding s_encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
 
-    public string? EncoderFallbackErrorMessage { get; init; }
-    public ImmutableArray<byte> FilePathChecksumOpt { get; init; }
-    public string DisplayFilePath { get; init; }
+    private string? _filePath;
 
-    public static FileIdentifier Create(SyntaxTree tree)
-        => Create(tree.FilePath);
+    private string? _encoderFallbackErrorMessage;
+    private string? _displayFilePath;
+    private ImmutableArray<byte> _filePathChecksumOpt;
 
-    public static FileIdentifier Create(string filePath)
+    private FileIdentifier(string filePath)
+    {
+        _filePath = filePath;
+    }
+
+    private FileIdentifier(ImmutableArray<byte> filePathChecksumOpt, string displayFilePath)
+    {
+        _filePathChecksumOpt = filePathChecksumOpt;
+        _displayFilePath = displayFilePath;
+    }
+
+    private void EnsureInitialized()
     {
-        string? encoderFallbackErrorMessage = null;
-        ImmutableArray<byte> hash = default;
-        try
+        var filePath = _filePath;
+        if (filePath is not null)
         {
-            var encodedFilePath = s_encoding.GetBytes(filePath);
-            using var hashAlgorithm = SourceHashAlgorithms.CreateDefaultInstance();
-            hash = hashAlgorithm.ComputeHash(encodedFilePath).ToImmutableArray();
+            string? encoderFallbackErrorMessage = null;
+            ImmutableArray<byte> hash = default;
+            try
+            {
+                var encodedFilePath = s_encoding.GetBytes(filePath);
+                using var hashAlgorithm = SourceHashAlgorithms.CreateDefaultInstance();
+                hash = hashAlgorithm.ComputeHash(encodedFilePath).ToImmutableArray();
+            }
+            catch (EncoderFallbackException ex)
+            {
+                encoderFallbackErrorMessage = ex.Message;
+            }
+
+            var displayFilePath = GeneratedNames.GetDisplayFilePath(filePath);
+
+            _encoderFallbackErrorMessage = encoderFallbackErrorMessage;
```

Comment:
I think we still don't have thread safety with this approach. I could be misunderstanding something, though. My suggestion was:
- make all the class fields `readonly`, and initialize their values in the constructor, rather than `EnsureInitialized`.
- for those types where we want to cache this, e.g. BuckStopsHereBinder, source type symbols, and PE type symbols, include a field of type `FileIdentifier?`, which initially has value `null`.
- When the accessor e.g. `AssociatedFileIdentifier` is called, and the field is null, interlocked-initialize the `FileIdentifier` field with a fully initialized instance and return it.

I think this most conveniently solves the thread safety questions while still giving us the perf win, as most of the binders and symbols will never actually create the instance.

I'm definitely open to other approaches and would like to make sure @333fred and @AlekseyTs would also be comfortable with this approach.

Reply:
I'm willing to go whatever approach you guys like best. I'm not clear on exactly where the concern lies with the current approach, and not clear on exactly what you are proposing, but if Fred and Aleksey like it better than what this PR has so far, I'm definitely willing to give that a go.

Reply:
Discussed offline with Aleksey and we think that what I've described is overkill, and also just moves the problem from internal to this type to external to it, in some ways.

I think it will be adequate to instead modify the properties like `DisplayFilePath`, etc. to `Volatile.Read` the fields, rather than just returning them normally.

The scenario I want to prevent is that we do not get in an inconsistent state, where another thread observes that `_filePath` is null, but also observes the "derived" fields like `_displayFilePath` as still null, due to inadequate synchronization.

Reply:
Instead of doing several Volatile.Reads, would it be sufficient to just do Volatile.Write(ref _filePath, null) when assigning _filePath instead?

Reply:
> Instead of doing several Volatile.Reads, would it be sufficient to just do Volatile.Write(ref _filePath, null) when assigning _filePath instead?

Yes, I think so. [Doc](https://learn.microsoft.com/en-us/dotnet/api/system.threading.volatile?view=net-7.0#remarks) says

> A volatile write operation prevents earlier memory operations on the thread from being reordered to occur after the volatile write.

Reply:
That's the way I read it too and it seemed easier. Just wanted to make sure I wasn't missing something.

Reply:
> A volatile write operation prevents earlier memory operations on the thread from being reordered to occur after the volatile write.

I believe Interlocked exchange provides the same guarantees. Also, this doesn't affect other threads, but @RikkiGibson was concerned about reads happening on another thread.